### PR TITLE
Expose `GetInputType` as a protected member of `InputTagHelper`.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Added doc comments.

#4979

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 1]
```diff
@@ -230,6 +230,33 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
+        /// </summary>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> to inspect.</param>
+        /// <param name="inputTypeHint">When this method returns, contains a type hint associated with the given
+        /// <paramref name="modelExplorer"/>.</param>
+        /// <returns>An &lt;input&gt; elements "type" attribute value.</returns>
+        /// <remarks><paramref name="inputTypeHint"/> values are based off of the <paramref name="modelExplorer"/>'s
+        /// <see cref="ModelExplorer.Metadata"/>. Input type hints are typically used to format an
+        /// &lt;input&gt; elements "value" attribute of a generated text box or used to determine if a text box should
+        /// be multi-line.</remarks>
+        protected string GetInputType(ModelExplorer modelExplorer, out string inputTypeHint)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: For the future, please do not move methods around while making other changes. It makes it harder to see what's really changing. We also aren't that persnickety about order in existing classes and someone could do a pass reordering class members on some future Engineering Day.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 2]
```diff
@@ -230,6 +230,33 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
             }
         }
 
+        /// <summary>
+        /// Gets an &lt;input&gt; elements "type" attribute value for the given <paramref name="modelExplorer"/>.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: No, `Gets an &lt;input&gt; element's "type" attribute value based the given <paramref name="modelExplorer"/> or <see cref="InputType"/>.` Note the apostrophe as well as the property mention.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 3]
```diff
@@ -230,6 +230,33 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
+        /// </summary>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> to inspect.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: No idea what you mean by "inspect". Suggest "use".

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 4]
```diff
@@ -230,6 +230,33 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
...
+        /// <param name="inputTypeHint">When this method returns, contains a type hint associated with the given
+        /// <paramref name="modelExplorer"/>.</param>
+        /// <returns>An &lt;input&gt; elements "type" attribute value.</returns>
+        /// <remarks><paramref name="inputTypeHint"/> values are based off of the <paramref name="modelExplorer"/>'s
+        /// <see cref="ModelExplorer.Metadata"/>. Input type hints are typically used to format an
+        /// &lt;input&gt; elements "value" attribute of a generated text box or used to determine if a text box should
+        /// be multi-line.</remarks>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Remove the `<remarks/>`. What you're saying here is misleading and not complete enough to be useful.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 5]
```diff
@@ -230,6 +230,33 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
             }
         }
 
+        /// <summary>
+        /// Gets an &lt;input&gt; elements "type" attribute value for the given <paramref name="modelExplorer"/>.
+        /// </summary>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> to inspect.</param>
+        /// <param name="inputTypeHint">When this method returns, contains a type hint associated with the given
+        /// <paramref name="modelExplorer"/>.</param>
+        /// <returns>An &lt;input&gt; elements "type" attribute value.</returns>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: "element's"

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 6]
```diff
@@ -230,6 +230,33 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
             }
         }
 
+        /// <summary>
+        /// Gets an &lt;input&gt; elements "type" attribute value for the given <paramref name="modelExplorer"/>.
+        /// </summary>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> to inspect.</param>
+        /// <param name="inputTypeHint">When this method returns, contains a type hint associated with the given
+        /// <paramref name="modelExplorer"/>.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: I wonder whether we should instead expose a `protected string GetInputType(ModelExplorer modelExplorer)` method. The `inputTypeHint` value is meaningless unless you're reading the heuristics in `GetInputTypeHints()` or how `GetFormat()` uses the string. Neither "type hint" nor "associated with the given ..." helps much.

If @marcusien or others find this value useful, it's closer to `... contains the string, often the name of a <see cref="ModelMetadata.ModelType"/> base class, used to determine this method's return value.`

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Curious, where does the InputType calculation come into play within this method?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: In the codez -- https://github.com/aspnet/Mvc/pull/5104/files/97a49f9044430e633d4f17675d621bd32b65dce5#diff-ed3977cf971ec6121e035dd0f173df3eR256

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs (hunk 6 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Haven't heard from @marcusien so I'm assuming more info is better than less here.

**Comment by NTaylorMullen** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 7]
```diff
@@ -230,6 +230,30 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
             }
         }
 
+        /// <summary>
+        /// Gets an &lt;input&gt; element's "type" attribute value based on the given <paramref name="modelExplorer"/>
+        /// or <see cref="InputType"/>.
+        /// </summary>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> to use.</param>
+        /// <param name="inputTypeHint">When this method returns, contains a type hint associated with the given
+        /// <paramref name="modelExplorer"/>.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Closer to `... contains the string, often the name of a <see cref="ModelMetadata.ModelType"/> base class, used to determine this method's return value.`

**Comment by dougbu** [comment]
:shipit: w/ one last comment on a doc comment to update the comment as I've commented.

**Pull request merged**